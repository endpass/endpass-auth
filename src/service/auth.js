// @ts-check
import get from 'lodash/get';
import request from '@/class/singleton/request';
import { AUTH_STATUS_CODE } from '@/constants';

const identityBaseUrl = ENV.VUE_APP_IDENTITY_API_URL;

const TIMEOUT_DEFAULT = 1500;

/**
 * @type {object}
 */
const CODE_TO_STATUS = {
  200: AUTH_STATUS_CODE.LOGGED_IN,
  401: AUTH_STATUS_CODE.NOT_LOGGED,
  403: AUTH_STATUS_CODE.NEED_PERMISSION,
};

/**
 * @param {Promise} promise
 * @return {Promise<any>}
 */
const withSuccess = promise => {
  return promise.then(res => {
    if (res && !res.success) throw new Error(res.message);

    return res;
  });
};

/**
 * @param {Function} handler
 * @return {number}
 */
const createTimeout = handler => setTimeout(handler, TIMEOUT_DEFAULT);

/**
 * @param {string} email
 * @return {Promise<any>}
 */
const getAuthChallenge = email => {
  return withSuccess(
    request.post(`${identityBaseUrl}/auth`, {
      email,
    }),
  );
};

/**
 * @param {string} password
 * @param {string} originHost
 * @return {Promise<any>}
 */
const setAuthPermission = async (password, originHost) => {
  const res = await request.post(`${identityBaseUrl}/auth/permission`, {
    password,
    originHost,
  });
  return res;
};

/**
 * @param {object} params
 * @param {string} params.email
 * @param {string} params.code
 * @param {string} params.password
 * @param {boolean} params.isSignUp
 * @param {boolean} params.isRemember
 * @return {Promise<any>}
 */
const authWithCode = async ({
  email,
  code,
  password,
  isSignUp,
  isRemember,
}) => {
  const url = isSignUp
    ? `${identityBaseUrl}/auth/signup`
    : `${identityBaseUrl}/auth/token`;

  return withSuccess(
    request.post(url, {
      email,
      code,
      password,
      remember: isRemember,
    }),
  );
};

/**
 * @param {string} idToken
 * @return {Promise<any>}
 */
const authWithGoogle = idToken =>
  request
    .post(`${identityBaseUrl}/auth/google`, {
      token: idToken,
    })
    .then(res => {
      if (!res.success) throw new Error(res.message);
      return res;
    })
    .catch(err => {
      throw err.response.data;
    });

/**
 * @param {string} code
 * @return {Promise<any>}
 */
const authWithGitHub = code =>
  request
    .post(`${identityBaseUrl}/auth/github`, {
      code,
    })
    .then(res => {
      if (!res.success) throw new Error(res.message);
      return res;
    })
    .catch(err => {
      throw err.response.data;
    });

const logout = () => request.post(`${identityBaseUrl}/logout`);

/**
 * @return {Promise<{expiresAt: number, status: string, hash: string}>}
 */
const getAuthStatus = async () => {
  try {
    const { expiresAt, hash } = await request.get(
      `${identityBaseUrl}/auth/check`,
    );
    return {
      status: AUTH_STATUS_CODE.LOGGED_IN,
      hash,
      expiresAt,
    };
  } catch (e) {
    const statusCode = get(e, ['response', 'status']);
    const status = CODE_TO_STATUS[statusCode] || AUTH_STATUS_CODE.NOT_LOGGED;

    const hash = get(e, ['response', 'data', 'hash'], '');
    const expiresAt = get(e, ['response', 'data', 'expiresAt'], 0);
    return {
      status,
      hash,
      expiresAt,
    };
  }
};

/**
 * @return {Promise<number>}
 */
const waitLogin = () =>
  new Promise((resolve, reject) => {
    /* eslint-disable-next-line */
    const handler = async function() {
      try {
        const { status } = await getAuthStatus();

        if (
          status === AUTH_STATUS_CODE.LOGGED_IN ||
          status === AUTH_STATUS_CODE.NEED_PERMISSION
        ) {
          return resolve();
        }

        createTimeout(handler);
      } catch (err) {
        return reject(err);
      }
    };

    return handler();
  });

/**
 * @param {string} email
 * @return {Promise<string>}
 */
const getRecoveryIdentifier = email =>
  request
    .get(`${identityBaseUrl}/auth/recover?email=${encodeURIComponent(email)}`)
    .then(res => {
      if (!res.success) throw new Error(res.message);

      return res.message;
    });

/**
 * @param {string} email
 * @return {Promise<number>}
 */
const sendEmailCode = async email => {
  const { timeout } = await request.post(`${identityBaseUrl}/auth/code`, {
    email,
  });

  return timeout;
};

/**
 * Send sms with code for disabling otp
 * @param {string} email
 * @returns {Promise<any>}
 */
const sendOtpRecoverSms = async email => {
  try {
    return withSuccess(
      request.get(
        `${identityBaseUrl}/auth/recover?email=${encodeURIComponent(email)}`,
      ),
    );
  } catch (error) {
    const { response = {} } = error;
    const { status } = response;
    error.code = status;

    throw error;
  }
};

/**
 * Disable otp setting
 * @param {object} param
 * @param {string} param.email
 * @param {string} param.code
 * @returns {Promise<any>}
 */
const disableOtpViaSms = async ({ email, code }) => {
  return withSuccess(
    request.post(`${identityBaseUrl}/auth/recover`, {
      email,
      code,
    }),
  );
};

/**
 * @return {string}
 */
const getSeedTemplateUrl = () => `${identityBaseUrl}/auth/recover/template`;

export default {
  getAuthStatus,
  setAuthPermission,
  getAuthChallenge,
  authWithCode,
  authWithGoogle,
  authWithGitHub,
  sendEmailCode,
  logout,
  waitLogin,
  getRecoveryIdentifier,
  sendOtpRecoverSms,
  disableOtpViaSms,
  getSeedTemplateUrl,
};
