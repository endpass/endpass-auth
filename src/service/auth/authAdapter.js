import { CHALLENGE_TYPES } from '@/constants';

function getChallengeType(challenge) {
  if (challenge === 'sms') {
    return CHALLENGE_TYPES.SMS_OTP;
  }

  if (challenge === 'otp') {
    return CHALLENGE_TYPES.APP_OTP;
  }

  return CHALLENGE_TYPES.EMAIL_OTP;
}

export default function(authData) {
  return {
    ...authData,
    challenge: {
      ...authData.challenge,
      challengeType: getChallengeType(authData.challenge),
    },
  };
}
