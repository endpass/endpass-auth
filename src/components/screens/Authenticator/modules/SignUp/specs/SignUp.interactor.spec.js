import Vuex from 'vuex';
import { shallowMount, createLocalVue } from '@vue/test-utils';
import { email, regularPassword as password } from '@unitFixtures/auth';
import SignUpInteractor from '@/components/screens/Authenticator/modules/SignUp/SignUp.interactor';
import SignUpView from '@/components/screens/Authenticator/modules/SignUp/SignUp.view';
import setupI18n from '@/locales/i18nSetup';
import authService from '@/service/auth';
import identityService from '@/service/identity';

const localVue = createLocalVue();
localVue.use(Vuex);
const i18n = setupI18n(localVue);

describe('SignUpInteractor', () => {
  let wrapper;
  const createWrapper = (options, props) =>
    shallowMount(SignUpInteractor, {
      localVue,
      propsData: {
        email,
        ...props,
      },
      i18n,
      ...options,
    });

  beforeEach(async () => {
    jest.clearAllMocks();

    wrapper = createWrapper();
    await global.flushPromises();
  });

  describe('render', () => {
    it('should correctly render component', () => {
      expect(wrapper.name()).toBe('SignUpInteractor');
      expect(wrapper.html()).toMatchSnapshot();
    });

    it('should render form', () => {
      expect(wrapper.find('sign-up-stub').exists()).toBe(true);
      expect(wrapper.html()).toMatchSnapshot();
    });
  });

  describe('behavior', () => {
    it('should emit sign-up event', () => {
      expect(wrapper.emitted()['sign-in']).toBeUndefined();

      wrapper.find(SignUpView).vm.$emit('sign-in');

      expect(wrapper.emitted()['sign-in'].length).toBe(1);
      expect(wrapper.emitted()['sign-in'][0]).toEqual([]);
    });

    describe('handle social event', () => {
      it('should call service', () => {
        expect(authService.waitLogin).not.toBeCalled();

        wrapper.find(SignUpView).vm.$emit('social');

        expect(authService.waitLogin).toBeCalledTimes(1);
        expect(authService.waitLogin).toBeCalledWith();
      });

      it('should emit event', async () => {
        expect.assertions(2);

        expect(wrapper.emitted().social).toBeUndefined();

        wrapper.find(SignUpView).vm.$emit('social');
        await global.flushPromises();

        expect(wrapper.emitted().social).toHaveLength(1);
      });

      describe('loading status', () => {
        it('should be false before event', () => {
          expect(
            wrapper.find('sign-up-stub').attributes().isloading,
          ).toBeFalsy();
        });

        it('should be true while waiting', () => {
          wrapper.find(SignUpView).vm.$emit('social');

          expect(wrapper.find('sign-up-stub').attributes().isloading).toBe(
            'true',
          );
        });

        it('should be false after sending', async () => {
          expect.assertions(1);

          wrapper.find(SignUpView).vm.$emit('social');
          await global.flushPromises();

          expect(
            wrapper.find('sign-up-stub').attributes().isloading,
          ).toBeFalsy();
        });
      });
    });

    describe('submit', () => {
      const defaultEventParams = {
        email,
        password,
      };

      describe('default auth mode', () => {
        it('should emit sign-in event', async () => {
          expect.assertions(3);

          const defaultParams = {
            ...defaultEventParams,
            isSignUp: true,
          };

          identityService.checkRegularPassword.mockResolvedValueOnce(true);

          expect(wrapper.emitted()['sign-up']).toBeUndefined();

          wrapper.find(SignUpView).vm.$emit('submit', defaultEventParams);
          await global.flushPromises();

          expect(wrapper.emitted()['sign-up'].length).toBe(1);
          expect(wrapper.emitted()['sign-up'][0]).toEqual([defaultParams]);
        });

        it('should get challenge type', () => {
          expect.assertions(3);

          expect(authService.getAuthChallenge).not.toBeCalled();

          wrapper.find(SignUpView).vm.$emit('submit', defaultEventParams);

          expect(authService.getAuthChallenge).toBeCalledTimes(1);
          expect(authService.getAuthChallenge).toBeCalledWith(email);
        });
      });

      it('should not handle submit event when loading status true', async () => {
        expect.assertions(1);

        wrapper.find(SignUpView).vm.$emit('submit', defaultEventParams);
        wrapper.find(SignUpView).vm.$emit('submit', defaultEventParams);
        await global.flushPromises();

        expect(wrapper.emitted()['sign-up'].length).toBe(1);
      });

      describe('loading status', () => {
        it('should be false before submit', () => {
          expect(
            wrapper.find('sign-up-stub').attributes().isloading,
          ).toBeFalsy();
        });

        it('should be true after submit', () => {
          wrapper.find(SignUpView).vm.$emit('submit', defaultEventParams);

          expect(wrapper.find('sign-up-stub').attributes().isloading).toBe(
            'true',
          );
        });

        it('should be false after handling submit', async () => {
          expect.assertions(1);

          wrapper.find(SignUpView).vm.$emit('submit', defaultEventParams);
          await global.flushPromises();

          expect(
            wrapper.find('sign-up-stub').attributes().isloading,
          ).toBeFalsy();
        });
      });

      describe('error', () => {
        beforeEach(() => {
          authService.getAuthChallenge.mockRejectedValueOnce(
            new Error('error'),
          );
        });

        it('should pass error', async () => {
          expect.assertions(2);

          expect(wrapper.find('sign-up-stub').attributes().error).toBeFalsy();

          wrapper.find(SignUpView).vm.$emit('submit', defaultEventParams);
          await global.flushPromises();

          expect(wrapper.find('sign-up-stub').attributes().error).toBe(
            i18n.t('components.compositeAuth.authFailed'),
          );
        });

        it('should remove error if exists before submit', async () => {
          expect.assertions(1);

          wrapper.find(SignUpView).vm.$emit('submit', defaultEventParams);
          await global.flushPromises();

          wrapper.find(SignUpView).vm.$emit('submit', defaultEventParams);

          expect(wrapper.find('sign-up-stub').attributes().error).toBeFalsy();
        });
      });
    });
  });
});
