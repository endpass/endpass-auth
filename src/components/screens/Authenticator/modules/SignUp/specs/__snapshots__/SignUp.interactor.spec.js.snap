// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUpInteractor render should correctly render component 1`] = `<sign-up-stub error="" email="user@email.com"></sign-up-stub>`;

exports[`SignUpInteractor render should render form 1`] = `<sign-up-stub error="" email="user@email.com"></sign-up-stub>`;
